source_ds9/mm_comms9.c

```diff
@@ -0,0 +1,500 @@
+    0,
+    3,
+    3,
+    0,
+    0,
+    0,
+    1,
+    2,
+    2,
+    2,
+    2,
+    2,
+    2,
+    0,
+    0,
+    1,
+    0,
+    3,
+    3,
+    0,
+    0,
+    3,
+    0,
+    0,
+    0,
+    0,
+    0,
+    1,
+    1,
+    0
+};
+
+static void SendSimpleExt(mm_hword, mm_byte, enum message_ids);
+static void SendSimple(mm_word, enum message_ids);
+static void SendString(mm_word*, int);
+static mm_sfxhand mmValidateEffectHandle(mm_sfxhand);
+static mm_sfxhand mmCreateEffectHandle(void);
+static void mmReceiveMessage(uint32_t, void*);
+
+// Fifo channel to use for communications
+mm_word mmFifoChannel;
+
+mm_word sfx_bitmask;
+
+mm_byte sfx_instances[EFFECT_CHANNELS];
+
+// ARM9 Communication Setup
+void mmSetupComms(mm_word channel) {
+    mmFifoChannel = channel;
+    fifoSetValue32Handler(channel, mmReceiveMessage, 0);
+}
+
+// Send data via Value32...?, wrapper
+static void SendSimpleExt(mm_hword data, mm_byte ext, enum message_ids msg_id) {
+    SendSimple(data | (ext << 24), msg_id);
+}
+
+// Send data via Value32...? Apparently, it was switched to Datamsg???
+static void SendSimple(mm_word data, enum message_ids msg_id) {
+    mm_word buffer[MAX_PARAM_WORDS];
+    mm_byte num_params = 0;
+    
+    // Determine number of parameters
+    if(msg_id < sizeof(message_num_params))
+        num_params = message_num_params[msg_id];
+    
+    // Store in buffer
+    buffer[0] = (data << 16) | (msg_id << 8) | (num_params + 1);
+    buffer[1] = data >> 24;
+    
+    // Send. If they're three bytes, do two transfers
+    if(num_params > 2)
+        SendString(buffer, 2);
+    else
+        SendString(buffer, 1);
+}
+
+// Send data via Datamsg
+static void SendString(mm_word* values, int num_words) {
+    fifoSendDatamsg(mmFifoChannel, num_words * sizeof(mm_word), (unsigned char*)values);
+}
+
+// Send a soundbank to ARM7
+void mmSendBank(mm_word num_songs, mm_addr bank_addr) {
+    mm_word buffer[MAX_PARAM_WORDS];
+    
+    buffer[0] = (num_songs << 16) | (MSG_BANK << 8) | 6;
+    buffer[1] = (mm_word)bank_addr;
+    
+    SendString(buffer, 2);
+}
+
+// Lock channels to prevent use by maxmod
+// TODO: Make this work for all channels, not just the PHYS ones.
+// Requires changing the receiving end in arm7 to use 5 bytes.
+void mmLockChannels(mm_word bitmask) {
+    SendSimpleExt(bitmask, 1, MSG_SELCHAN);
+}
+
+// Unlock channels to allow use by maxmod
+// TODO: Make this work for all channels, not just the PHYS ones.
+// Requires changing the receiving end in arm7 to use 5 bytes.
+void mmUnlockChannels(mm_word bitmask) {
+    SendSimpleExt(bitmask, 0, MSG_SELCHAN);
+}
+
+// Start module playback
+void mmStart(mm_word module_ID, mm_pmode mode) {
+    mmActiveStatus = 1;
+    SendSimpleExt(module_ID, mode, MSG_START);
+}
+
+// Pause module playback
+void mmPause() {
+    SendSimple(0, MSG_PAUSE);
+}
+
+// Resume module playback
+void mmResume() {
+    SendSimple(0, MSG_RESUME);
+}
+
+// Stop module playback
+void mmStop() {
```

Comment:
Don't use this kind of prototypes without arguments. Use ``(void)`` in all functions instead of ``()``.

Reply:
I agree! And when I make prototypes, you can see that I do use void (unless I make a mistake).
However, this function's prototype is actually here: https://github.com/blocksds/maxmod/blob/fe4d5ac802c94d4d498bef5b09287076c7cfc73b/include/maxmod9.h#L210

Reply:
Yeah, but then we should be consistent and not use different parameter names in the implementation and the declaration, or different signatures.

Reply:
Made signatures use void as well.